### DISCUSSION TITLE
`--textmode` and EOLs

### DISCUSSION THREAD

**Discussion by bjansen** [pgpainless/pgpainless#191] [Q&A] [answered]
Hi,

I'm looking to switch my current project from bouncy-gpg to pgpainless, mostly because bouncy-gpg is kinda dormant and has bugs that forced us to maintain a fork.

I'm trying to assess whether pgpainless fixes all our problems. The main one is a proper support of an equivalent to gpg's `--textmode` flag. From what I understand of [the RFC](https://datatracker.ietf.org/doc/html/rfc4880), it's a combination of the following:
* input data is treated as text (duh)
* EOLs have to be normalized to `\r\n`, which means potentially converting EOLs *before* signing the message (see [section 5.9](https://datatracker.ietf.org/doc/html/rfc4880#section-5.9))
* some flags need to be set to mark literal data packet as `t`ext

Our main problem with bouncy-gpg is that the current version does not support `PGPSignature.CANONICAL_TEXT_DOCUMENT` yet (I opened a PR but the maintainer seems to be busy). 
pgpainless supports it though, so it's cool. Except there seem to be another [bug in bcpg](https://stackoverflow.com/a/40155151/500478), which _does_ canonicalize text before computing the signature, but _does not_ use this canonicalized data in literal data packets. As a result, when we send our encrypted files to another system, this system fails to check the signature (it's a mainframe app that uses some IBM PGP implementation I believe). 

So, my question is: does pgpainless 'fix' this bug in bcpg, i.e. does it transform line endings before encrypting/signing data when `SignatureType.CANONICAL_TEXT_DOCUMENT` is used?

**Comment by vanitasvitae** [comment]
You raise an interesting issue that I wasn't aware of yet.
It appears that PGPainless does in fact not fix this issue. I will create some tests and try to upstream a fix.

In the mean time I will see if I can come up with a workaround for PGPainless.

**Reply by bjansen** [reply]
The workaround we have currently is basically `content.replaceAll("\\r?\\n", "\r\n")` and then feed that to bouncy-gpg.

**Reply by vanitasvitae** [reply]
For completeness: I reported the issue upstream: https://github.com/bcgit/bc-java/issues/1035
I also started working on a possible upstream fix: https://github.com/pgpainless/bc-java/commit/470fb6068b94d7f13fb75ed244599345b6364d89

**Comment by vanitasvitae** [comment, marked as answer]
I created an upstream PR fixing this issue: https://github.com/bcgit/bc-java/pull/1040

**Comment by bjansen** [comment]
So, when BC 1.70 and its `PGPCanonicalizedDataGenerator` are released, how will that fit into PGPainless? Will you make a new release that uses this class, or will applications have to use it directly?

**Reply by bjansen** [reply]
Ah, there's already work in progress, didn't see it :) https://github.com/pgpainless/pgpainless/commit/1539e1da123ab59d0984ae18797e03fb97eb38af

**Reply by vanitasvitae** [reply]
Indeed, PGPainless will transparently make use of the newly introduced `PGPCanonicalizedDataGenerator` without requiring the user to do anything :)

**Reply by vanitasvitae** [reply]
Just for the record, it appears that there are some [issues](https://github.com/pgpainless/pgpainless/issues/264) with mixed literal data formats / signature types.

**Comment by vanitasvitae** [comment]
I added some remarks to [#264](https://github.com/pgpainless/pgpainless/issues/264#issuecomment-1083206738) stating my motivation behind reverting PGPainless' behavior on CRLF encoding.

**Reply by vanitasvitae** [reply]
Update: In #265 I added another method `ProducerOptions.applyCRLFEncoding()` which can be used to automatically apply CRLF encoding.

**Comment by bjansen** [comment]
Hey @vanitasvitae, I just wanted to let you know that we've successfully migrated from bouncy-gpg to pgpainless in our application. Thanks again for your fixes, everything works great for our needs now!

Just a quick question though. We _have_ to use these two deprecated options, otherwise we can't read files encrypted by our z/OS mainframe (or the other way around, z/OS can't read files generated by our Java app):

* `ConsumerOptions.setIgnoreMDCErrors(true)`
* `ProducerOptions.setEncoding(StreamEncoding.TEXT)`

Are those two methods deprecated because they're bad practice, or because you intend to completely remove them in a future version?

**Reply by vanitasvitae** [reply]
`setIgnoreMDCErrors()` is bad practice, as it enables consuming encrypted data packets which are not integrity protected. This opens up some possible attacks (see [efail](https://efail.de/)).

`setEncoding()` is deprecated simply because it is advised that applications leave this field at the default value, however it will stay around as it doesn't do harm either.

Thank you for letting me know that you rely on these functions though!
I'm glad PGPainless fits your needs :D
